language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -220,47 +242,86 @@ typeOf (TyInst x body ty) = do
             k' <- kindOf ty
             typeCheckStep
             if k == k'
-                then pure (tyReduce (tySubstitute (extractUnique n) (void $ NormalizedType ty) (NormalizedType absTy)))
+                then do
+                    tyEnvAssign (extractUnique n) (void $ NormalizedType ty)
+                    tyReduce absTy <* tyEnvDelete (extractUnique n)
                 else throwError (KindMismatch x (void ty) k k')
         _ -> throwError (TypeMismatch x (void body) (TyForall () dummyTyName dummyKind dummyType) nBodyTy)
 typeOf (Unwrap x body) = do
     nBodyTy@(NormalizedType bodyTy) <- typeOf body
     case bodyTy of
         TyFix _ n fixTy -> do
-            let subst = tySubstitute (extractUnique n) nBodyTy (NormalizedType fixTy)
-            pure (tyReduce subst)
+            tyEnvAssign (extractUnique n) nBodyTy
+            tyReduce fixTy <* tyEnvDelete (extractUnique n)
```

Comment:
These two lines are repeated above, probably worth making this a named function. `tyReduceBinder` or something like that.